acme_tiny.py

```diff
@@ -104,7 +104,7 @@ def _send_signed_request(url, payload):
             raise ValueError("Error requesting challenges: {0} {1}".format(code, result))
 
         # make the challenge file
-        challenge = [c for c in json.loads(result.decode('utf8'))['challenges'] if c['type'] == "http-01"][0]
+        challenge = [c for c in json.loads(result.decode('utf8'))['challenges'] if c['type'] == "http-01" or c['status'] == "valid"][0]
```

Comment:
What do you try to achieve here? This is collecting (potentially) more challenges than there have been before, and it treats them all as `http-01` challenges, which is wrong.

I think what you *want* to do is to first take a look at `json.loads(result.decode('utf8'))`'s field `status`; if that is `'valid'`, skip creating the challenge file.

Reply:
Please see my pull request comments for an explanation. This changed works for me.

Reply:
I read your comments. The thing is that the authz object is valid, but the concrete challenges might not be. Your code does *not* check whether the authz object is valid (it is ignored as before), but you do something which happens to work in your particular case but which does *not* work in general. If the order of the challenges returned is different, your fix might not work at all, because it picks the `http-01` challenge instead of the valid one (which can show up later in the list).

Reply:
(The authz object is `json.loads(result.decode('utf8'))`.)

Reply:
I disagree with your analysis. If there is an http-01 challenge to pick first, then it *should* be picked first and that would be fine. In the specific case of working against the letsencrypt.org server, the server will return either the type == http-01 (the case that is typically being looked for), or status == valid (this case of mine with an authorization previously obtained by a different mechanism or possibly some other case where status == valid) or some other case acme-tiny doesn't handle. Trying to make this code more general is, IMO, a waste of effort.

Reply:
The server returns a list of challenges, which has no well-defined order, and you always pick the first challenge satisfying one of the conditions `type == 'http-01' or status == 'valid'`. That does *not* work in general.

Reply:
IMO, this code only needs to handle the specific cases encountered by users of acme-tiny in conjunction with letsencrypt.org. In my use, I never have seen (or can imagine) a case where the order mattered.